jax/tree_util.py

```diff
@@ -247,7 +274,23 @@ def _replace_nones(sentinel, tree):
       return tree
 
 no_initializer = object()
-def tree_reduce(function, tree, initializer=no_initializer):
+
+# TODO: Introduce overloads that verify the initializer type when all callers
+# are okay with them.
+# @overload
+# def tree_reduce(function: Callable[[T, Any], T],
+#                 tree: Any) -> T:
+#     ...
+#
+# @overload
+# def tree_reduce(function: Callable[[T, Any], T],
+#                 tree: Any,
+#                 initializer: T) -> T:
+#     ...
+
+def tree_reduce(function: Callable[[T, Any], T],
```

Comment:
You'll actually have to remove the type annotation completely for me to submit this. (It's the output type causing the problem, so you could also just change the output type to `Any` for now.)

That said, we think this annotation caught a real bug in a RL agent, so it's also fine to reinstate the overloads if you don't mind waiting another day to get this in... Up to you.

Reply:
Oh, great, happy to hear the type annotations caught something! :smile: 

Let's wait then, so that we can do a more perfect job.

Reply:
I think I might be able to get this in now. Please push a version with the overloads added back!

Reply:
Done!